posts/2024-05-01-spring-boot-3.adoc

```diff
@@ -0,0 +1,232 @@
++
+  Hello from Spring Boot in Open Liberty!
++
+8. When you finish testing the application, stop the server by typing `<Ctrl>+C`.
++
+9. Recap
++
+The server is configured with the https://openliberty.io/docs/latest/reference/feature/springBoot-3.0.html[springBoot-3.0] feature. The liberty-maven-plugin has been added to the pom.xml along with special configuration to use the optimized Spring Boot deployment.  The app is packaged as an "executable WAR" by running the `spring-boot:repackage` goal in the `package` phase, and the app is bootstrapped via its `main()` method in `SpringLibertyApplication` passing in application arguments defined in `server.xml`.
++
+If you're wondering, yes, while a Spring Boot WAR must be repackaged as an executable WAR to use the optimized deployment, that executable WAR could still be deployed as a standard WAR.
+
+=== Build container image with efficient layering
+
+Now that we have used the optimized Spring Boot deployment, we can efficiently build a container image.  This image uses an indexed cache at `/lib.index.cache` to store Spring Boot dependencies in their own layer, separate from your application code.
+
+1. Create your `Dockerfile`.  
++
+[source,dockerfile]
+----
+# Stage and thin the application 
+FROM icr.io/appcafe/open-liberty:full-java21-openj9-ubi-minimal as staging
+
+ARG APPNAME=spring-liberty-0.0.1-SNAPSHOT.war
+COPY --chown=1001:0 target/$APPNAME \
+  /staging/$APPNAME
+
+RUN springBootUtility thin \
+ --sourceAppPath=/staging/$APPNAME \
+ --targetThinAppPath=/staging/thin-$APPNAME \
+ --targetLibCachePath=/staging/lib.index.cache
+
+FROM icr.io/appcafe/open-liberty:kernel-slim-java21-openj9-ubi-minimal
+
+ARG APPNAME=spring-liberty-0.0.1-SNAPSHOT.war
+ARG VERSION=1.0
+ARG REVISION=SNAPSHOT
+COPY --chown=1001:0 src/main/liberty/config/server.xml /config/server.xml
+
+RUN features.sh
+
+COPY --chown=1001:0 --from=staging /staging/lib.index.cache /lib.index.cache
+COPY --chown=1001:0 --from=staging /staging/thin-$APPNAME \
+                    /config/apps/thin-$APPNAME
+
+RUN configure.sh 
+----
++
+(Note you can use the `full-java17-openj9-ubi` tag to build the equivalent Java 17 image.)
++
+2. Build then run the image
++
+[source,sh]
+----
+docker build -t springboot:demo .
+docker run -p 9080:9080 -p 9443:9443 -it springboot:demo
+----
+
+== Results
+
+To recap, a Spring Boot WAR can be deployed to Liberty like any other WAR, or it can be deployed with an optimized deployment using special liberty-maven-plugin configuration and the `springBoot-3.0` feature configured in server.xml.
+Though much of the programming model is the same across the two cases, there are some differences including the bootstrap mechanism.
+
+== References
+* Clone the https://github.com/scottkurz/spring-liberty[repository] with the finished code for this blog.
+* doc:   https://openliberty.io/docs/latest/deploy-spring-boot.html   
+* instanton: https://openliberty.io/blog/2023/09/26/spring-boot-3-instant-on.html
```

Comment:
instantOn isn't explained or mentioned in the post, I'd remove this unless you add some context